scripts/set_assignees.py

```diff
@@ -90,30 +91,114 @@ def process_pr(gh, maintainer_file, number):
+
+        #
+        # Not sure what the right percentage is. When a PR has touched
+        # a lot of files it will create a giant list. So try to find
+        # just the top areas touched and hopefully it will get the right
+        # labels.
+        #
+        prop = (area_counter[item] / num_files) * 100
+        # log(f":::: label {item.labels} : count {area_counter[item]} : {prop:.2f}%")
+
+        if prop > 15:
+            for l in item.labels:
+                if l not in labels:
+                    labels.append(l)
+
     log(f"Area matches: {area_counter}")
     log(f"labels: {labels}")
 
     # Create a list of collaborators ordered by the area match
+    #
+    # There is a limit to how many reviewers can be assigned
+    # to the PR. The following allocation strategy attempts
+    # to more fairly distribute the assigned reviewers across
+    # the impacted areas.
+    #
+    # Bias walking through each touched area to pull in the
+    # maintainers from that area into the collab list
+    # first. Then process the collaborators by picking one
+    # collaborator at a time from each area impacted. Note
+    # that the order of the names in each collaborator list
+    # will matter if there is a large array of areas being
+    # touched as it will be more likely to select someone
+    # from the top of the list. As such, it may be more
+    # prudent to put higher priority people at the top of
+    # the list... sort of a "promotion ladder". Also, when an
+	# area's collaboration list is exhausted, it is dropped
+	# from the iterator list.
+    #
+	# Exammple:
+    # Areas (by priority/impact order): A, B, C, D
+    #
+    # collab list would look like:
+    #
+    # maintainers [A, B, C, D] followed by
+    # collaborators [A1, B1, C1, D1, A2, B2, C2, D2, A3, ... ]
+    #
+	# If there is only 1 collaborator in area B then the array is:
+	#
+	# collaborators [A1, B1, C1, D1, A2, C2, D2, A3, ... ]
+    #
+    # An alternative is to just walk the collabs list taking them
+    # in groups by just doing the loop again:
+    #
+    #     for area in area_counter:
+    #         collab += maintainer_file.areas[area.name].maintainers
+    #
+    # collaborators [A1, A2, A3, B1, B2, C1, C2, D1, D2, ... ]
+    #
+    # Down side seem to be a strong bias to the top list, which
+    # may be better.
+    #
     collab = list()
     for area in area_counter:
         collab += maintainer_file.areas[area.name].maintainers
-        collab += maintainer_file.areas[area.name].collaborators
+
+    iterators = {area.name: iter(maintainer_file.areas[area.name].collaborators) for area in area_counter}
+    while iterators:
+        for key in list(iterators):
+            try:
+                if iterators[key] not in collab:
+                    collab.append(next(iterators[key]))
+                else:
+                    next(iterators[key])
+            except StopIteration:
+                del iterators[key]
```

Comment:
Maybe doing something like the following is easier to understand for the future maintainers, what do you think?

```py
areas = ['bluetooth', 'drivers']
    area_reviewers = {
        'bluetooth': {
            'maintainers': ['jori', 'alwa'],
            'collaborators': ['jori', 'thb1', 'alwa', 'vich'],
        },
        'drivers': {
            'maintainers': ['oth2'],
            'collaborators': ['oth1', 'oth3', 'vich', 'oth5'],
        },
    }

    reviewers = list()
    # add all maintainers for all the affected areas
    for area in areas:
        reviewers += area_reviewers[area]['maintainers']

    # loop over all the areas, adding one collaborator at a time
    # exit when all the collaborators have been added
    while True:
        added_someone = False
        for area in areas:
            for collaborator in area_reviewers[area]['collaborators']:
                if collaborator not in reviewers:
                    added_someone = True
                    reviewers.append(collaborator)

                    break # go to the next area

        # when there is no more collaborator to add → exit
        if not added_someone:
            break

    expected_reviewers = ['jori', 'alwa', 'oth2', 'thb1', 'oth1', 'vich', 'oth3', 'oth5']

    assert reviewers == expected_reviewers
```

Reply:
you don't like my fancy iterator ;-)

Let me digest this later. Thanks.

Reply:
@theob-pro after playing with yours I realize it that it still has a bias where earlier areas may get over represented in the reviewer's list from later areas. You have a loop that continues looping through each collaborator in an area until it can add one to the collaborator list which then allows it to move on to the next area. Because we have maintainers that are also collaborators, this style of looping will allow an area to have potentially an over-representation of an area in the beginning of the review list. 

When pulling from collaborators, I was trying to construct a selection algorithm that would walk through each area and select the first collaborator in the list and then move to the next area to grab one. This is why I have a list of area iterators. The "while" loop is to keep looping the inner loop while there are iterators in the list. The inner loop just looks at each iterator and pulls the first name from that area's collab list. If that name is already in the list it will move on to the next area iterator. Note that this means the area already has representation either from a someone who is a collaborator in another area or a maintainer... but it still means that area has reviewer representation so move to the next area. 

If the area iterator has no more names (or never had any), it is removed from the iterator list and the loop proceeds to the next area. Eventually, each iterator in the list will be exhausted and removed from the list so that the 'while iterators:' loop can eventually exit.

Reply:
@theob-pro I changed my algorithm to use deque() types and perform a round-robin merge of the area collaborator lists into the collab list (which is what I was doing before but with an array of iterators). I think this is easier to understand. Let me know.